Step2 - 모델링 하기

sjhello opened the pull request:
안녕하세요 리뷰어님.
1단계에서 피드백 주신 `속성과 행위를 분리`하라는 피드백에 조금 응용을 하여 모델링을 하면서 `정책도 분리`를 해보았습니다.

하실 말씀이 있으시다면 편하게 말씀해주세요! 감사합니다 :)

sjhello committed:
docs: 상품 모델링 추가, 상품 용어사전 수정

sjhello committed:
docs: 메뉴 그룹 모델링 추가 및 용어사전 수정, 상품 모델링, 용어사전 수정

sjhello committed:
docs: 메뉴 그룹 모델링 추가

sjhello committed:
docs: 주문 테이블 모델링 추가

sjhello committed:
docs: 상품, 메뉴 그룹, 메뉴 모델링 수정

 - 메뉴 용어 사전 수정

sjhello committed:
docs: 주문 모델링 추가 및 용어사전 수정

sjhello committed:
refactor: 어색한 표현 수정, 단위 추가

sjhello committed:
docs: 주문 테이블 용어 사전 수정

kyucumber commented on README.md:
```diff
@@ -183,37 +188,197 @@ docker compose -p kitchenpos up -d
+- `Product`는 `Product Price`를 갖는다.
+- `Product`는 `Product Name`을 갖는다.
+
+#### 행위
+- `Product`를 등록한다.
+- `Product Price`를 변경한다.
+- `Product` 목록을 조회한다.
+
+#### 정책
+- `Product Price`는 음수가 될 수 없다.
+- `Product Name`는 `Profanity`를 포함 할 수 없다.
+- `Produc Price`가 변경될 때 `Menu Product` 상품 금액의 합보다 크다면 메뉴가 숨겨진다.
```
정책을 따로 나누는것도 좋지만 명확하게 연결된 속성과 행위가 존재하는 경우라면 같이 묶어서 정리되는게 더 가독성 및 유지보수성이 향상될 것 같습니다.

속성에 대해서만 보고 넘어가서 정책을 이해하지 못하거나, 속성을 수정하는 경우 아래 정책에 반영하는걸 누락하거나 할 수 있을 것 같아서요.

sjhello replied:
속성과 직접적으로 연결되어지는 `유효성 검사 종류`의 정책은 속성과 함께 적는 것으로 수정 하겠습니다.

kyucumber commented on README.md:
```diff
@@ -183,37 +188,197 @@ docker compose -p kitchenpos up -d
 ## 모델링
+
+### 상품
+#### 속성
```
꼭 필요한 부분은 아니나 속성의 특성에 따라 데이터 타입과 관련되어 논의가 필요한 경우가 생길 수 있을 것 같아요. 모델링 과정에서 일부 속성의 경우(금액은 BigDecimal 등)은 데이터 타입을 명시해도 좋을 것 같습니다.

sjhello replied:
모델링에 대해 작성하면서 `키친 포스`를 프로그램을 사용하였을때에 `어떠한 일이 일어나는지?`에 대해 집중하였습니다. 이에 대해 데이터 타입은 중요하지 않은 정보라 생각하였기 때문에 작성하지 않았습니다. 또한 `BigDecimal` 등의 데이터 타입이 `개발자 친화적`이라고 생각이 듭니다. 때문에 모델링에서 데이터 타입은 명시하지 않아도 괜찮을 것 같습니다.

kyucumber replied:
저는 모델링 결과로 개발을 진행하는데 이런 타입이나 영문명을 명확하게 해두지 않으면 결국 모델링한 걸 기반으로 다시 논의를 해야해서 불편하더라구요. 개발자 친화적인 언어로 적히는게 조금 불편하다면 소수점 둘째자리까지를 포함한 최대 20자리 수 등으로 풀어서 적을수도 있을 것 같네요.

중요한 부분은 아니라 답변만 남겨두고 Resolve 하도록 할게요. 참고 부탁드립니다.

kyucumber commented on README.md:
```diff
@@ -183,37 +188,197 @@ docker compose -p kitchenpos up -d
+- `Menu`는 식별자를 갖는다.
+- `Menu`는 `Menu Name`을 갖는다.
+- `Menu`는 `Menu Price`를 갖는다.
+- `Menu`는 `Menu Group`을 갖는다.
```
필수적인 부분은 아니지만, 모델링이므로 도메인 개체 간(메뉴, 메뉴 그룹, 메뉴 상품 등)의 연관 관계 및 제약 조건 등을 다이어그램으로 그리거나 1 : N 등으로 표현해보면 어떨까요?

```suggestion
- `Menu`는 `Menu Group`을 갖는다. (1 : N)
```

kyucumber commented on README.md:
```diff
@@ -183,37 +188,197 @@ docker compose -p kitchenpos up -d
+- `Menu Price`는 `Menu Product`의 가격(`상품의 가격` * `상품의 수량`) 보다 같거나 낮아야 한다.
+
+##### `Menu`의 공개 상태를 `공개`로 변경
+- `Menu Price`가 `Menu Product`의 가격(`상품의 가격` * `상품의 수량`)이 높은 경우 `Menu`를 `공개`로 변경하지 않는다.
```
변경하지 않는게 아니라 공개할 수 없다고 표현되는게 맞지 않을까요?

sjhello replied:
그렇네요 `변경하지 않는다`보다는 `공개 할 수 없다`가 자연스러운 표현 같습니다. :)

kyucumber reviewed (changes requested):
모델링 잘 정리해주셨네요!

간단한 몇가지 코멘트 남겨두었으니 확인 부탁드려요.

sjhello committed:
refactor: 어색한 표현 수정

sjhello committed:
refactor: 피드백 적용

 - 단순한 유효성 검사 종류의 정책을 속성으로 옮긴다.
 - `Menu`, `Menu Group`, `Menu Product`의 연관관계를 표현한다.

kyucumber reviewed (approved):
모델링 잘 마무리해주셨네요.

간단한 코멘트 하나 남겨두었으니 참고 부탁드려요.

https://github.com/next-step/ddd-strategic-design/pull/415#discussion_r1617308842

다음 단계 진행을 위해 머지하겠습니다.
